src/linux/namespaces.rs

```diff
@@ -0,0 +1,335 @@
+//! Linux namespaces.
+
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::ffi::{CStr, CString};
+use std::fs::{self, File};
+use std::io::Error as IoError;
+use std::os::unix::ffi::OsStrExt;
+use std::path::{Component, Path, PathBuf};
+use std::{env, ptr};
+
+use bitflags::bitflags;
+
+use crate::error::Result;
+
+/// Path for mount namespace's new root.
+const NEW_ROOT: &str = "/tmp/birdcage-root";
+
+/// Old root mount point inside the new root.
+///
+/// This should not conflict with any files or directories which might be
+/// present in a root directory, since it will be placed in the new root.
+///
+/// However this is only present within the new bind mount, so it will not be
+/// persisted and cannot conflict with previous mount namespace sandboxes.
+const OLD_ROOT_DIR: &str = "birdcage-old-root";
+
+/// Isolate process using Linux namespaces.
+///
+/// If successful, this will always clear the abstract namespace.
+///
+/// Additionally it will isolate network access if `deny_networking` is `true`.
+pub fn create_namespaces(
+    allow_networking: bool,
+    bind_mounts: HashMap<PathBuf, libc::c_ulong>,
+) -> Result<()> {
+    // Get EUID/EGID outside of the namespace.
+    let uid = unsafe { libc::geteuid() };
+    let gid = unsafe { libc::getegid() };
+
+    // Setup the network namespace.
+    if !allow_networking {
+        create_user_namespace(0, 0, Namespaces::NETWORK)?;
+    }
+
+    // Isolate filesystem and procfs.
+    create_mount_namespace(bind_mounts)?;
+
+    // Drop root user mapping and ensure abstract namespace is cleared.
+    create_user_namespace(uid, gid, Namespaces::empty())?;
+
+    Ok(())
+}
+
+/// Create a mount namespace to isolate filesystem access.
+///
+/// This will deny access to any path which isn't part of `bind_mounts`. Allowed
+/// paths are mounted according to their bind mount flags.
+fn create_mount_namespace(bind_mounts: HashMap<PathBuf, libc::c_ulong>) -> Result<()> {
+    // Create mount namespace to allow creation of new mounts.
+    create_user_namespace(0, 0, Namespaces::MOUNT)?;
+
+    // Get target paths for new and old root.
+    let new_root = PathBuf::from(NEW_ROOT);
+    let put_old = new_root.join(OLD_ROOT_DIR);
+
+    // Ensure new root is available as an empty directory.
+    if new_root.exists() {
+        fs::remove_dir_all(&new_root)?;
+    }
+    fs::create_dir(&new_root)?;
```

Comment:
Hmm originally I thought this wouldn't be a problem since the directory is "invisible" to the outside anyway without the bind mounts. But since I had to remove it to clean up the old directory trees, it would also pull out those directories for existing sandboxes.

Maybe we have to go back to generating a new one for every sandbox. Should definitely consider cleanup options at that point though.